rfc/rfc-37/rfc-37.md

```diff
@@ -0,0 +1,286 @@
+            "doc": "Contains information about column ranges for all data files in the table",
+            "type": [
+                "null",
+                {
+                    "type": "record",
+                    "name": "HoodieColumnStats",
+                    "fields": [
+                        {
+                            "name": "rangeLow",
+                            "type": [
+                                "null",
+                                "string"
+                            ],
+                            "doc": "Low end of the range. For now, this is a String. Based on main data table schema, we can convert it to appropriate type"
+                        },
+                        {
+                            "name": "rangeHigh",
+                            "type": [
+                                "null",
+                                "string"
+                            ],
+                            "doc": "High end of the range. For now, this is a String. Based on main data table schema, we can convert it to appropriate type"
+                        },
+                        {
+                            "name": "isDeleted",
+                            "type": "boolean",
+                            "doc": "True if this file has been deleted"
+                        }
+                    ]
+                }
+            ]
+        }
+```
+
+Column stats records hold key ranges (min and max) for the file. The key for the column stat record would be an
+encoded string representing the tuple set of column name, partition name and the base file. The string names of
+these fields are converted to deterministic hash based IDs, and then they are base64 encoded, just like the
+bloom filter key.
+
+```
+key = base64_encode(hash64(column name) + hash64(partition name) + hash128(file path))
+```
+
+While Hash based IDs have quite a few desirable properties in the context of Hudi index lookups, there is an impact
+on the column level schema changes though. Refer to [Schema Evolution](#Schema-Evolution) section for more details.
+
+Below picture gives a pictorial representation of Column stats partition in metadata table.
+<img src="metadata_index_col_stats.png" alt="Column Stats Partition" width="480"/>
+
+### Metadata Index lookup:
+
+For the incoming upsert records, given their keys, tag their current location. The new algorithm for the
+index lookup would be
+
+1. Generate the list of partitions and the list of keys under each partition to be looked up
+2. For all the involved partitions, load all its file list
+3. Level 1: Range pruning using `column_stats` index:
+   1. For each of the record key, generate the column stats index lookup key based on the tuple 
+      (__hoodie_record_key, partition name, file path)
+   2. Meta index lookup with the above key and if available get the value payload with the column stats details
+   3. Build the key range interval tree from all the looked up keys
```

Comment:
I don't think we actually need this step since each individual key will have its own resolution sequence that will have to do Index Lookups (w/o reading it out whole in memory)